core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/pythonworker/PythonProxyServer.scala

```diff
@@ -0,0 +1,113 @@
+package edu.uci.ics.amber.engine.architecture.pythonworker
+
+import edu.uci.ics.amber.engine.architecture.messaginglayer.{ControlOutputPort, DataOutputPort}
+import edu.uci.ics.amber.engine.common.ambermessage.InvocationConvertUtils.{
+  controlInvocationToV1,
+  returnInvocationToV1
+}
+import edu.uci.ics.amber.engine.common.ambermessage._
+import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import kotlin.NotImplementedError
+import org.apache.arrow.flight._
+import org.apache.arrow.flight.example.InMemoryStore
+import org.apache.arrow.memory.{BufferAllocator, RootAllocator}
+import org.apache.arrow.util.AutoCloseables
+
+import scala.collection.mutable
+
+private class AmberProducer(
+    allocator: BufferAllocator,
+    location: Location,
+    controlOutputPort: ControlOutputPort,
+    dataOutputPort: DataOutputPort
+) extends InMemoryStore(allocator, location) {
+
+  override def doAction(
+      context: FlightProducer.CallContext,
+      action: Action,
+      listener: FlightProducer.StreamListener[Result]
+  ): Unit = {
+    action.getType match {
+      case "control" =>
+        val pythonControlMessage = PythonControlMessage.parseFrom(action.getBody)
+        pythonControlMessage.payload match {
+          case returnInvocation: ReturnInvocationV2 =>
+            controlOutputPort.sendTo(
+              to = pythonControlMessage.tag,
+              payload = returnInvocationToV1(returnInvocation)
+            )
+
+          case controlInvocation: ControlInvocationV2 =>
+            controlOutputPort.sendTo(
+              to = pythonControlMessage.tag,
+              payload = controlInvocationToV1(controlInvocation)
+            )
+
+        }
+        listener.onNext(new Result("ack".getBytes))
+        listener.onCompleted()
+      case _ => throw new NotImplementedError()
+    }
+
+  }
+
+  override def acceptPut(
+      context: FlightProducer.CallContext,
+      flightStream: FlightStream,
+      ackStream: FlightProducer.StreamListener[PutResult]
+  ): Runnable = { () =>
+    val dataHeader: PythonDataHeader = PythonDataHeader
+      .parseFrom(flightStream.getDescriptor.getCommand)
+    val to: ActorVirtualIdentity = dataHeader.tag
+    val isEnd: Boolean = dataHeader.isEnd
+
+    val root = flightStream.getRoot
+    try {
+      // consume all data in the stream, it will store on the root vectors.
+      while (flightStream.next) {
+        ackStream.onNext(PutResult.metadata(flightStream.getLatestMetadata))
+      }
+      // closing the stream will release the dictionaries
+      flightStream.takeDictionaryOwnership
+
+      if (isEnd) {
+        // EndOfUpstream
+        dataOutputPort.sendTo(to, EndOfUpstream())
```

Comment:
can we add an assert rowcount == 0, just to be safe in case we change the `isEnd` semantics in the future

Reply:
added